### PR TITLE
Fix flaky tests

### PR DESCRIPTION
**Pull request by ryanbogan**
### Description
Fixes the flaky tests on k-NN
 
### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/ModelIT.java [hunk 1]
```diff
@@ -140,13 +144,22 @@ public void testKNNModelDefault() throws Exception {
     }
 
     // KNN Delete Model test for model in Training State
+    @Ignore
```
Comment (heemin32) [inline qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/ModelIT.java:147]: Isn't this disabling the unit test?

File: src/test/java/org/opensearch/knn/plugin/transport/GetModelResponseTests.java [hunk 2]
```diff
@@ -23,8 +27,22 @@
...
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.mockStatic;
+import static org.mockito.Mockito.when;
+
 public class GetModelResponseTests extends KNNTestCase {
 
+    private static MockedStatic<KNNClusterUtil> knnClusterUtilMockedStatic;
+
+    @BeforeClass
+    public static void setup() {
```
Comment (heemin32) [inline src/test/java/org/opensearch/knn/plugin/transport/GetModelResponseTests.java:39]: Could you tell why this is related with flakiness of tests?

File: qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/ModelIT.java [hunk 3]
```diff
@@ -140,13 +144,22 @@ public void testKNNModelDefault() throws Exception {
...
             addModelToSystemIndex(TEST_MODEL_ID_TRAINING, testModelMetadata, testModelBlob);
-        } else {
+
+            Response getResponse = getModel(TEST_MODEL_ID_TRAINING, null);
+            String responseBody = EntityUtils.toString(getResponse.getEntity());
+            assertNotNull(responseBody);
+
+            Map<String, Object> responseMap = createParser(MediaTypeRegistry.getDefaultMediaType().xContent(), responseBody).map();
+            assertEquals(TEST_MODEL_ID_TRAINING, responseMap.get(MODEL_ID));
+            assertEquals(ModelState.TRAINING.toString(), responseMap.get(MODEL_STATE));
```
Comment (heemin32) [inline qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/ModelIT.java:161]: The unit test is for testing delete training model. However, why do we set state and verify the value which doesn't seems related with delete training model.

File: qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/ModelIT.java (hunk 1 above)
Comment (ryanbogan) [reply qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/ModelIT.java:147]: Yes, we are temporarily disabling this bwc test until we can create one that does not rely on the bug that I fixed

File: src/test/java/org/opensearch/knn/plugin/transport/GetModelResponseTests.java (hunk 2 above)
Comment (ryanbogan) [reply src/test/java/org/opensearch/knn/plugin/transport/GetModelResponseTests.java:39]: For some reason, the getClusterMinVersion() method in KNNClusterUtil returns a random version, not the most recent.  So the unit test would fail when the random version was before 2.12 or 3.0, since the field added wasn't in those.  This change ensures that the current version for each branch is returned.

File: qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/ModelIT.java (hunk 3 above)
Comment (ryanbogan) [reply qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/ModelIT.java:161]: We set the state so that the actual training job doesn't run, the model is just added as immediately in the "training" state.  Verifying the value is not required for the test, and was not removed since the test is currently ignored.

File: src/test/java/org/opensearch/knn/plugin/transport/GetModelResponseTests.java (hunk 2 above)
Comment (heemin32) [reply src/test/java/org/opensearch/knn/plugin/transport/GetModelResponseTests.java:39]: The randomness of the version might be an intended behavior. I am afraid your code here change that behavior for other test cases as well.

We need to check if there are any other tests which rely on randomness of the version. Also, changing behavior of static method should be done for entire test level but not in individual test level(Maybe inside KNNTestCase class?) so that there won't be any race condition among tests.

File: qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/ModelIT.java (hunk 1 above)
Comment (heemin32) [reply qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/ModelIT.java:146]: Can we just go ahead and write a correct test? I think marking test as ignored is not a good idea. If we really need to ignore it temporary, we should create a github issue to track the task to fix it.

File: qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/ModelIT.java [hunk 4]
```diff
@@ -140,13 +143,14 @@ public void testKNNModelDefault() throws Exception {
...
-        byte[] testModelBlob = "hello".getBytes(StandardCharsets.UTF_8);
-        ModelMetadata testModelMetadata = getModelMetadata();
-        testModelMetadata.setState(ModelState.TRAINING);
         if (isRunningAgainstOldCluster()) {
+            byte[] testModelBlob = "hello".getBytes(StandardCharsets.UTF_8);
+            ModelMetadata testModelMetadata = getModelMetadata();
+            testModelMetadata.setState(ModelState.TRAINING);
             addModelToSystemIndex(TEST_MODEL_ID_TRAINING, testModelMetadata, testModelBlob);
-        } else {
+
```
Comment (heemin32) [inline qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/ModelIT.java:153]: Could you tell why this change is needed? Especially on why else statement is not required?

Comment (ryanbogan) [reply qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/ModelIT.java:153]: The previous test relied on the bug where models stuck in training state would still be in training state when the cluster restarts.  Since I modified that, the test is no longer working.

File: src/test/java/org/opensearch/knn/plugin/transport/GetModelResponseTests.java (hunk 2 above)
Comment (ryanbogan) [reply src/test/java/org/opensearch/knn/plugin/transport/GetModelResponseTests.java:39]: Wouldn't the variable be localized to the class, unless it was specifically accessed elsewhere?

I was trying to only make this change for the classes that required a version check, so that any intended use for the randomized version would stay the same for the rest of the tests.

File: qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/ModelIT.java (hunk 1 above)
Comment (ryanbogan) [reply qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/ModelIT.java:146]: I can create a github issue.  I tried creating a new correct test, but the outcome was not what I expected so it will take more time.  For now, I think it's best to either ignore or delete the test temporarily so that we can not have any flaky tests.

Comment (heemin32) [reply qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/ModelIT.java:146]: The test is going to be removed in this PR. https://github.com/opensearch-project/k-NN/pull/1307
I would suggest to remote the test code here as same and create an github issue to add working test code later.

File: qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/ModelIT.java (hunk 4 above)
Comment (heemin32) [reply qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/ModelIT.java:153]: I am still don't' understand why we need a test for only old cluster. Anyway, this test need to be rewritten so we can stop the further discussion of this matter.

File: src/test/java/org/opensearch/knn/plugin/transport/GetModelResponseTests.java (hunk 2 above)
Comment (heemin32) [reply src/test/java/org/opensearch/knn/plugin/transport/GetModelResponseTests.java:39]: You are using static mock which won't be scoped to your test automatically.
https://www.baeldung.com/mockito-mock-static-methods

File: qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/ModelIT.java (hunk 4 above)
Comment (ryanbogan) [reply qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/ModelIT.java:153]: I agree, it should probably be a normal integration test instead of bwc

File: qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/ModelIT.java (hunk 1 above)
Comment (ryanbogan) [reply qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/ModelIT.java]: Looks like we already have a normal integ test for attempting to delete a training model, so I'm not sure why the bwc test exists too: https://github.com/opensearch-project/k-NN/blob/main/src/test/java/org/opensearch/knn/plugin/action/RestDeleteModelHandlerIT.java#L83

I believe that test and this github issue (https://github.com/opensearch-project/k-NN/issues/1365) will suffice to cover testing this functionality

File: src/test/java/org/opensearch/knn/plugin/transport/GetModelResponseTests.java (hunk 2 above)
Comment (ryanbogan) [reply src/test/java/org/opensearch/knn/plugin/transport/GetModelResponseTests.java:39]: I'll attempt to use the try method suggested in that article

**Review by heemin32 — APPROVED** [review]

File: qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/ModelIT.java [hunk 5]
```diff
@@ -55,7 +53,7 @@ public class ModelIT extends AbstractRestartUpgradeTestCase {
     private static int DOC_ID_TEST_MODEL_INDEX = 0;
     private static int DOC_ID_TEST_MODEL_INDEX_DEFAULT = 0;
     private static final int DELAY_MILLI_SEC = 1000;
-    private static final int EXP_NUM_OF_MODELS = 3;
+    private static final int EXP_NUM_OF_MODELS = 2;
```
Comment (martin-gaievski) [inline qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/ModelIT.java:56]: did you find out why expected num of models has changed from 3 to 2?

Comment (ryanbogan) [reply qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/ModelIT.java:56]: Yes, it's because there was another model created in the testDeleteTrainingModel() method that was deleted in this PR.  The remaining tests only create two models.

**Review by martin-gaievski — APPROVED** [review]

**Comment by martin-gaievski** [comment]
@ryanbogan - I assume we need this change in 2.x too?

**Comment by ryanbogan** [comment]
Yeah, this should be backported

**Comment by martin-gaievski** [comment]
> Yeah, this should be backported

Nice, wasn't sure that you can add PR labels
